prow/spyglass/podlogartifact.go

```diff
@@ -0,0 +1,167 @@
+)
+
+// NewPodLogArtifact creates a new PodLogArtifact
+func NewPodLogArtifact(jobName string, buildID string, podName string, sizeLimit int64, ja *jobs.JobAgent) (*PodLogArtifact, error) {
+	if jobName == "" {
+		return &PodLogArtifact{}, errInsufficientJobInfo
+	}
+	if buildID == "" {
+		return &PodLogArtifact{}, errInsufficientJobInfo
+	}
+	return &PodLogArtifact{
+		name:      jobName,
+		buildID:   buildID,
+		podName:   podName,
+		sizeLimit: sizeLimit,
+		ja:        ja,
+	}, nil
+}
+
+// CanonicalLink returns a link to where pod logs are streamed
+func (a *PodLogArtifact) CanonicalLink() string {
+	q := url.Values{
+		"job": []string{a.name},
+		"id":  []string{a.buildID},
+	}
+	u := url.URL{
+		Path:     "/log",
+		RawQuery: q.Encode(),
+	}
+	return u.String()
+}
+
+// JobPath gets the path within the job for the pod log. Always returns build-log.txt.
+// This is because the pod log becomes the build log after the job artifact uploads
+// are complete, which should be used instead of the pod log.
+func (a *PodLogArtifact) JobPath() string {
+	return "build-log.txt"
+}
+
+// ReadAt implements reading a range of bytes from the pod logs endpoint
+func (a *PodLogArtifact) ReadAt(p []byte, off int64) (n int, err error) {
+	logs := a.fetchLogs()
+	return bytes.NewReader(logs).ReadAt(p, off)
+}
+
+// ReadAll reads all available pod logs, failing if they are too large
+func (a *PodLogArtifact) ReadAll() ([]byte, error) {
+	size, err := a.Size()
+	if err != nil {
+		return []byte{}, nil
+	}
+	if size > a.sizeLimit {
+		return []byte{}, viewers.ErrFileTooLarge
+	}
+	return a.fetchLogs(), nil
+}
+
+// ReadAtMost reads at most n bytes
+func (a *PodLogArtifact) ReadAtMost(n int64) ([]byte, error) {
+	logs := a.fetchLogs()
+	reader := bytes.NewReader(logs)
+	var byteCount int64
+	var p []byte
+	for byteCount <= n {
+		b, err := reader.ReadByte()
+		if err != nil {
+			if err == io.EOF {
+				return p, nil
+			}
+			return p, err
+		}
+		p = append(p, b)
+		byteCount++
+	}
+	return p, nil
+}
+
+// ReadTail reads the last n bytes of the pod log
+func (a *PodLogArtifact) ReadTail(n int64) ([]byte, error) {
+	logs, err := a.ja.GetJobLogTail(a.name, a.buildID, n)
+	if err != nil {
+		logrus.WithField("artifactName", a.name).WithError(err).Error("Error getting pod logs tail")
+	}
+	off := int64(len(logs)) - n - 1
+	p := []byte{}
+	_, err = bytes.NewReader(logs).ReadAt(p, off)
+	if err != nil {
+		logrus.WithError(err).WithField("artifactName", a.name).Error("Failed to read pod logs.")
+	}
+	return p, err
+}
+
+// Size gets the size of the pod log. Note: this function makes the same network call as reading the entire file.
+func (a *PodLogArtifact) Size() (int64, error) {
+	var logs []byte
+	var err error
+	logs, err = a.ja.GetJobLog(a.name, a.buildID)
+	if err != nil {
+		return 0, err
+	}
+	return int64(len(logs)), nil
+
+}
+
+// UseContext is unsupported for pod logs
+func (a *PodLogArtifact) UseContext(ctx context.Context) error {
```

Comment:
A more natural pattern would be for `NewPodLogArtifact()` to accept a `context.Context` rather than have a `UseContext()`

Reply:
The reason for `UseContext` is that the behavior of the context needs to be specified by the viewers, since they are the ones making the read calls. The viewers dont construct artifacts themselves, only receive a list of `Artifact`.

Reply:
Also, if we specify context in the `Read` methods, we lose conformance to `io.ReaderAt`. I discussed this in Steve's (?) reviews a bit more in depth.

Reply:
It isn't clear to me why `Artifact` needs a `UseContext` interface instead of just ensuring that whatever creates the artifacts creates it with a context.

Is there any chance that multiple viewers read the same artifact? If so then that won't work with UseContext anyway.

Reply:
`Artifacts` are created by the `ArtifactFetcher`. If we create them with a context there and dont have `UseContext`, there's no way for the `viewers` to use the context in any meaningful way. There is a chance of two viewers reading the same artifact, I'd imagine it would actually be quite common. Would it work to have a `WithContext` method that returns a new `Artifact` instead of mutating the current one?

Reply:
It's not essential, but it was part of an earlier discussion about inserting context so that viewers could cancel read requests